vowpalwabbit/parse_example.cc

```diff
@@ -4,17 +4,45 @@
 
 #include <cmath>
 #include <cctype>
+#include <future>
+#include <thread>
+#include <chrono>
 #include "parse_example.h"
 #include "hash.h"
 #include "unique_sort.h"
 #include "global_data.h"
 #include "constant.h"
 #include "vw_string_view.h"
+#include "io_to_queue.h"
+
+class IO_Item;
 
 size_t read_features(vw* all, char*& line, size_t& num_chars)
 {
-  line = nullptr;
-  size_t num_chars_initial = readto(*(all->p->input), line, '\n');
+ 
+  IO_Item result;
+
+  while(true){
```

Comment:
Can you explain why this needs to be in a loop?

Reply:
Since the I/O thread and parser thread are running in parallel, there are moments in the execution of the program when the io queue stored in IO_State is empty, but read_features and the parser should still be expecting lines of code to parse. For example, this occurs occasionally when the I/O thread and parsing of lines first begins, when read_features must wait until an input line is added to the io queue before it can move on to parse and process it. The while loop enforces this waiting when a line is still expected from the io queue. 

Is it ok to keep this while loop, or is there another design I should be considering instead?

Reply:
This should be done via condition variable instead of a busy loop